mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp

```diff
@@ -0,0 +1,594 @@
+      }
+    }
+    toCommonReductionDim.merge("d_nr", dimInsertionPoint++, nonReduceDimNames);
+    toCommonReductionDim.merge("d_r", dimInsertionPoint++, reduceDimNames);
+  }
+  TransformMapAttr commonReduction = toCommonReductionDim.get();
+  LLVM_DEBUG(llvm::dbgs() << "commonReduction = " << commonReduction << "\n");
+  auto toResplitReduction =
+      BottomUpTMBuilder::above(toCommonReductionDim, commonReduction);
+  {
+    unsigned upperDimCount = commonReduction.getUpperBounds().size();
+    int64_t commonReduceSize =
+        commonReduction.getUpperBounds().asArrayRef().back();
+    int64_t commonFactor = math_util::gcd(commonReduceSize, dPerBlock);
+    toResplitReduction.passThrough(ArrayRef<unsigned>{0, 1, 3},
+                                   ArrayRef<unsigned>{0, 1, 2});
+    toResplitReduction.unmerge({"d_rh", "d_rl"}, {2, upperDimCount}, "d_r",
+                               {commonReduceSize / commonFactor, commonFactor});
+  }
+  TransformMapAttr resplitReduction = toResplitReduction.get();
+  LLVM_DEBUG(llvm::dbgs() << "resplitReductionAttr = " << resplitReduction
+                          << "\n");
+  auto toRecombined =
+      BottomUpTMBuilder::above(toResplitReduction, resplitReduction);
+  {
+    toRecombined.passThrough(ArrayRef<unsigned>{0, 1},
+                             ArrayRef<unsigned>{0, 1});
+    toRecombined.merge("d", 2, {"d_rh", "d_nr", "d_rl"});
+  }
+  TransformMapAttr recombined = toRecombined.get();
+  LLVM_DEBUG(llvm::dbgs() << "recombined = " << recombined << "\n");
+  OpBuilder builder(recombined.getContext());
+  return builder.getArrayAttr(
+      {reduceSplit, commonReduction, resplitReduction, recombined});
+}
+
+std::tuple<ArrayAttr, ArrayAttr> generateShuffledGemmInputViews(
+    OpBuilder &builder, int64_t g, int64_t m, int64_t mPerBlock, int64_t k,
+    int64_t n, int64_t nPerBlock,
+    llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>> reductionSubDims) {
+  BottomUpTMBuilder toReductionSplitA(builder, {"G", "K", "M"}, {g, k, m});
+  ArrayAttr additionalViewsA = builder.getArrayAttr({});
+  if (!reductionSubDims[1].empty()) {
+    additionalViewsA = reorderReductionDims(
+        toReductionSplitA, reductionSubDims[1], "M", m, mPerBlock);
+  }
+  BottomUpTMBuilder toReductionSplitB(builder, {"G", "K", "N"}, {g, k, n});
+  ArrayAttr additionalViewsB = builder.getArrayAttr({});
+  if (!reductionSubDims[2].empty()) {
+    additionalViewsB = reorderReductionDims(
+        toReductionSplitB, reductionSubDims[2], "N", n, nPerBlock);
+  }
+  return {additionalViewsA, additionalViewsB};
+}
+
+ArrayAttr generateShuffledGemmOutputViews(
+    OpBuilder &builder, int64_t g, int64_t m, int64_t mPerBlock, int64_t n,
+    int64_t nPerBlock,
+    llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>> reductionSubDims) {
+  // Split the reduction and non-reduction splits
+  int64_t totalReductionSizeM = 1;
+  for (const SubDimInfo &sdInfo : reductionSubDims[1]) {
```

Comment:
we should not assume reductionSubDims contains 1 or 2

Reply:
We only shuffle dimensions where we do tiling and G is not a tiling dimension.
I ll add a comment.